motorboard/mbed-firmware/main.cpp

```diff
@@ -334,11 +341,13 @@ void pid() {
     ErrorL = desiredSpeedL - actualSpeedL;
     ErrorR = desiredSpeedR - actualSpeedR;
 
-    dErrorL = ErrorL - lastErrorL;
-    dErrorR = ErrorR - lastErrorR;
+    // dPV/dt
+    dErrorL = (actual_speed_last_l - actualSpeedL)/dT_sec;
```

Comment:
shouldn't this be desiredSpeedL?

Reply:
`desiredSpeedL - actualSpeed` yields `ErrorL`. I think what you mean is why it isn't `ErrorL - lastErrorL`. The reason why it's not `dError/dt` but rather `dPV/dt` is because when `SP` changes, `dError/dt` rises to almost infinity, which leads to weird behavior, ie. when we start from standstill, or when we execute a turn, and one of the motors' setpoint changes. Instead, since `Error = SP - PV`, when `SP = 0`, then `Error = -PV`, and so `dError/dt = -dPV/dt`.

Reply:
:+1: